### PR TITLE
feat: search users

### LINKED ISSUE #20: Proxy request to search users
and create an audit log record with the identity of the requester

### PR DESCRIPTION
**Pull request by xcoulon**
controller that proxies to `auth`, and records
and audit log for the operation

fixes #20

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: .make/test.mk [hunk 1]
```diff
@@ -429,6 +425,22 @@ $(COV_PATH_INTEGRATION): $(SOURCES) $(GOCOVMERGE_BIN)
...
+$(MINIMOCK_BIN):
+	@echo "building the minimock binary..."
+	@cd $(VENDOR_DIR)/github.com/gojuno/minimock/cmd/minimock && go build -v minimock.go
+
+.PHONY: generate-mocks
+## Generate mocks
+generate-mocks: deps $(MINIMOCK_BIN) 
+	@echo "Generating mocks..."
+	-mkdir -p test/generated/configuration
+	$(MINIMOCK_BIN) -i vendor/github.com/fabric8-services/fabric8-common/token.ManagerConfiguration -o ./test/generated/configuration/manager_configuration_mock.go -t ManagerConfigurationMock
```
Comment (alexeykazakov) [inline .make/test.mk]: Add test/generated to clean target as well?

File: Gopkg.toml [hunk 2]
```diff
@@ -68,3 +69,7 @@ ignored = [
 [[constraint]]
   name = "github.com/jteeuwen/go-bindata"
   revision = "v3.0.7"
+
+[[constraint]]
+  name = "github.com/gojuno/minimock"
+  branch = "master"
```
Comment (alexeykazakov) [inline Gopkg.toml]: Don't relay on the latest master. Add revision.

File: configuration/defaults.go [hunk 3]
```diff
@@ -6,4 +6,36 @@ const (
 	defaultDBPassword = "mysecretpassword"
 
 	defaultLogLevel = "info"
+
+	// DevModeRsaPrivateKey for signing JWT Tokens in Dev Mode
+	// ssh-keygen -f alm_rsa
+	DevModeRsaPrivateKey = `-----BEGIN RSA PRIVATE KEY-----
```
Comment (alexeykazakov) [inline configuration/defaults.go]: Just reuse what we already have in https://github.com/fabric8-services/fabric8-common/blob/master/configuration/configuration.go#L41

File: controller/search.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	if claims, ok := tk.Claims.(jwt.MapClaims); ok {
+		var err error
+		sub := claims["sub"].(string)
+		identityID, err = uuid.FromString(sub)
+		if err != nil {
+			return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("invalid authorization token (invalid 'sub' claim)"))
+		}
+	} else {
+		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("invalid authorization token (invalid type of claims)"))
+	}
```
Comment (alexeykazakov) [inline controller/search.go]: https://github.com/fabric8-services/fabric8-common/blob/master/token/token.go#L199 does what you need here (lines 50-65)

File: test/jwt/jwt_context.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"context"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	goajwt "github.com/goadesign/goa/middleware/security/jwt"
+)
+
+// NewJWTContext creates a context with a JWT having the given `subject`
+// using the private key that is located in the 'privateKeyPath'
+func NewJWTContext(subject, kid, privateKey string) (context.Context, error) {
```
Comment (alexeykazakov) [inline test/jwt/jwt_context.go]: Not sure why you need multiple keys in your tests.. Why not just using https://github.com/fabric8-services/fabric8-common/blob/master/test/auth/auth.go#L119?

File: main.go [hunk 6]
```diff
@@ -109,12 +112,22 @@ func main() {
 			metric.WithRequestDurationBucket(prometheus.ExponentialBuckets(0.05, 2, 8))))
 	service.WithLogger(goalogrus.New(log.Logger()))
 
-	// service.Use(metric.Recorder())
+	appDB := application.NewGormApplication(db)
+	tokenParser, err := token.NewManager(config)
```
Comment (alexeykazakov) [inline main.go]: I would recommend using `token.DefaultManager(config)` instead. Then you can access the token manager later even if you don't have incoming request.

File: .make/test.mk (hunk 1 above)
Comment (xcoulon) [reply .make/test.mk]: ok, let me add that

File: Gopkg.toml (hunk 2 above)
Comment (xcoulon) [reply Gopkg.toml]: ok

File: configuration/defaults.go (hunk 3 above)
Comment (xcoulon) [reply configuration/defaults.go]: ah yes! thanks!

File: controller/search.go (hunk 4 above)
Comment (xcoulon) [reply controller/search.go]: ok, thanks for pointing this!

File: main.go (hunk 6 above)
Comment (xcoulon) [reply main.go]: yes, that's what I'll need anyways, so I can use the `Locate` method to extract the `sub` claim from the request token ;)

File: test/jwt/jwt_context.go (hunk 5 above)
Comment (xcoulon) [reply test/jwt/jwt_context.go]: because one of my test uses an invalid key (signed by another party) to verify that the request is not authorized

Comment (xcoulon) [reply test/jwt/jwt_context.go]: On second thought, the token signature validation is the responsibility of the goa middleware which is invoked before the controller is called, so testing with different signing keys does not make sense at his level.

**Comment by xcoulon** [comment]
@alexeykazakov I've addressed all your comments. Please let me know if you have further concerns with regards to this PR.

File: main.go [hunk 7]
```diff
@@ -109,12 +113,29 @@ func main() {
 			metric.WithRequestDurationBucket(prometheus.ExponentialBuckets(0.05, 2, 8))))
 	service.WithLogger(goalogrus.New(log.Logger()))
 
-	// service.Use(metric.Recorder())
+	appDB := application.NewGormApplication(db)
+	tokenManager, err := token.NewManager(config)
```
Comment (alexeykazakov) [inline main.go]: Still would use `token.DefaultManager(config)` instead :) https://github.com/fabric8-services/fabric8-common/blob/master/token/token.go#L43
Both options will work but it's a good practice to initiate the default manager in main().
In this case you will have the same token manager instance used in all request contexts and get it available even if you don't have any context at all (for example if you need to start some goroutine to do some backround work and you need a token manager for something).

See example here:
https://github.com/fabric8-services/fabric8-cluster/blob/master/main.go#L152

Comment (xcoulon) [reply main.go]: ok, let me apply that change

**Pull request merged**